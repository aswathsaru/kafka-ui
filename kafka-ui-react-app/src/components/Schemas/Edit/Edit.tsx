import Breadcrumb from 'components/common/Breadcrumb/Breadcrumb';
import { NewSchemaSubject, SchemaSubject } from 'generated-sources';
import { clusterSchemaPath, clusterSchemasPath } from 'lib/paths';
import React from 'react';
import { ClusterName, SchemaName } from 'redux/interfaces';
import PageLoader from 'components/common/PageLoader/PageLoader';
import { useHistory } from 'react-router';
import JSONEditor from 'components/common/JSONEditor/JSONEditor';

export interface EditProps {
  subject: SchemaName;
  schema: SchemaSubject;
  clusterName: ClusterName;
  schemasAreFetched: boolean;
  createSchema: (
    clusterName: ClusterName,
    newSchemaSubject: NewSchemaSubject
  ) => Promise<void>;
  fetchSchemasByClusterName: (clusterName: ClusterName) => void;
}

export default function Edit({
  subject,
  schema,
  clusterName,
  schemasAreFetched,
  createSchema,
  fetchSchemasByClusterName,
}: EditProps) {
  let newSchema = '';

  React.useEffect(() => {
    if (!schemasAreFetched) fetchSchemasByClusterName(clusterName);
  }, [clusterName, fetchSchemasByClusterName]);

  const handleChange = (e: string) => {
    newSchema = e;
  };

  const history = useHistory();
  const onSubmit = async () => {
    try {
      await createSchema(clusterName, {
        ...schema,
        schema: newSchema,
      });
      history.push(clusterSchemaPath(clusterName, subject));
    } catch (e) {
      // Show Error
    }
  };
  return (
    <div className="section">
      <div className="level">
        <div className="level-item level-left">
          <Breadcrumb
            links={[
              {
                href: clusterSchemasPath(clusterName),
                label: 'Schema Registry',
              },
              {
                href: clusterSchemaPath(clusterName, subject),
                label: subject,
              },
            ]}
          >
            Edit
          </Breadcrumb>
        </div>
      </div>

      {schemasAreFetched ? (
        <div className="box">
          <div className="is-flex is-justify-content-space-evenly">
            <div>
              <h4 className="title is-5 mb-2">Latest Schema</h4>
              <JSONEditor
                readonly
                value={JSON.stringify(JSON.parse(schema.schema), null, '\t')}
                name="latestSchema"
              />
            </div>
            <div>
              <h4 className="title is-5 mb-2">New Schema</h4>
              <JSONEditor
                value={newSchema}
                name="newSchema"
                onChange={(e) => handleChange(e)}
              />
            </div>
          </div>
          <div className="is-flex is-justify-content-center is-align-items-center">
            <button
              type="submit"
              className="button is-primary mt-3"
              onClick={onSubmit}
            >
              Submit
            </button>
          </div>
        </div>
      ) : (
        <PageLoader />
      )}
    </div>
  );
}
