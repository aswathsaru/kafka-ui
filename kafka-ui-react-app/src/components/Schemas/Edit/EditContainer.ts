import { connect } from 'react-redux';
import { ClusterName, RootState } from 'redux/interfaces';
import { RouteComponentProps, withRouter } from 'react-router-dom';
import {
  getIsSchemaListFetched,
  getSchema,
} from 'redux/reducers/schemas/selectors';
import { createSchema, fetchSchemasByClusterName } from 'redux/actions';
import Edit from './Edit';

interface RouteProps {
  clusterName: ClusterName;
  subject: string;
}

type OwnProps = RouteComponentProps<RouteProps>;

const mapStateToProps = (
  state: RootState,
  {
    match: {
      params: { clusterName, subject },
    },
  }: OwnProps
) => ({
  subject,
  schema: getSchema(state, subject),
  schemasAreFetched: getIsSchemaListFetched(state),
  clusterName,
});

const mapDispatchToProps = {
  createSchema,
  fetchSchemasByClusterName,
};

export default withRouter(connect(mapStateToProps, mapDispatchToProps)(Edit));
