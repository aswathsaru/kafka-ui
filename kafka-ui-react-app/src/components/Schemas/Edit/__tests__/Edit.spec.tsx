import { shallow } from 'enzyme';
import { SchemaType } from 'generated-sources';
import React from 'react';
import Edit from '../Edit';

describe('Edit Component', () => {
  const mockSchema = {
    subject: 'Subject',
    version: '1',
    id: 1,
    schema: '{"schema": "schema"}',
    compatibilityLevel: 'BACKWARD',
    schemaType: SchemaType.AVRO,
  };

  describe('when schemas are not fetched', () => {
    const component = shallow(
      <Edit
        subject="Subject"
        clusterName="ClusterName"
        schemasAreFetched={false}
        fetchSchemasByClusterName={jest.fn()}
        createSchema={jest.fn()}
        schema={mockSchema}
      />
    );
    it('matches the snapshot', () => {
      expect(component).toMatchSnapshot();
    });
    it('shows loader', () => {
      expect(component.find('PageLoader').exists()).toBeTruthy();
    });
  });

  describe('when schemas are fetched', () => {
    const component = shallow(
      <Edit
        subject="Subject"
        clusterName="ClusterName"
        schemasAreFetched
        fetchSchemasByClusterName={jest.fn()}
        createSchema={jest.fn()}
        schema={mockSchema}
      />
    );
    it('matches the snapshot', () => {
      expect(component).toMatchSnapshot();
    });
    it('shows editor', () => {
      expect(component.find('JSONEditor').length).toEqual(2);
      expect(component.find('button').exists()).toBeTruthy();
    });
    it('calls createSchema on button click', () => {
      const mockCreateSchema = jest.fn();
      const componentWithMockFn = shallow(
        <Edit
          subject="Subject"
          clusterName="ClusterName"
          schemasAreFetched
          fetchSchemasByClusterName={jest.fn()}
          createSchema={mockCreateSchema}
          schema={mockSchema}
        />
      );
      componentWithMockFn.find('button').simulate('click');
      expect(mockCreateSchema).toHaveBeenCalledWith('ClusterName', {
        ...mockSchema,
        schema: '',
      });
      expect(mockCreateSchema).toHaveBeenCalledTimes(1);
    });
  });
});
