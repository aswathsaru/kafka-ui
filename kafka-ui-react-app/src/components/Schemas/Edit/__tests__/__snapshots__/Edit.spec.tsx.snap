// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Edit Component when schemas are fetched matches the snapshot 1`] = `
<div
  className="section"
>
  <div
    className="level"
  >
    <div
      className="level-item level-left"
    >
      <Breadcrumb
        links={
          Array [
            Object {
              "href": "/ui/clusters/ClusterName/schemas",
              "label": "Schema Registry",
            },
            Object {
              "href": "/ui/clusters/ClusterName/schemas/Subject/latest",
              "label": "Subject",
            },
          ]
        }
      >
        Edit
      </Breadcrumb>
    </div>
  </div>
  <div
    className="box"
  >
    <div
      className="is-flex is-justify-content-space-evenly"
    >
      <div>
        <h4
          className="title is-5 mb-2"
        >
          Latest Schema
        </h4>
        <JSONEditor
          name="latestSchema"
          readonly={true}
          value="{
	\\"schema\\": \\"schema\\"
}"
        />
      </div>
      <div>
        <h4
          className="title is-5 mb-2"
        >
          New Schema
        </h4>
        <JSONEditor
          name="newSchema"
          onChange={[Function]}
          value=""
        />
      </div>
    </div>
    <div
      className="is-flex is-justify-content-center is-align-items-center"
    >
      <button
        className="button is-primary mt-3"
        onClick={[Function]}
        type="submit"
      >
        Submit
      </button>
    </div>
  </div>
</div>
`;

exports[`Edit Component when schemas are not fetched matches the snapshot 1`] = `
<div
  className="section"
>
  <div
    className="level"
  >
    <div
      className="level-item level-left"
    >
      <Breadcrumb
        links={
          Array [
            Object {
              "href": "/ui/clusters/ClusterName/schemas",
              "label": "Schema Registry",
            },
            Object {
              "href": "/ui/clusters/ClusterName/schemas/Subject/latest",
              "label": "Subject",
            },
          ]
        }
      >
        Edit
      </Breadcrumb>
    </div>
  </div>
  <PageLoader />
</div>
`;
